compositor_pipeline/src/error.rs

```diff
@@ -113,34 +117,94 @@ impl From<UnregisterOutputError> for PipelineError {
     }
 }
 
-// placeholders for now
+const FAILED_TO_CREATE_NODE: &str = "FAILED_TO_CREATE_NODE";
+const SCENE_SPEC_VALIDATION_ERROR: &str = "SCENE_SPEC_VALIDATION_ERROR";
+const MISSING_NODE_WITH_ID: &str = "MISSING_NODE_WITH_ID";
+const UNKNOWN_RESOLUTION_ON_OUTPUT_NODE: &str = "UNKNOWN_RESOLUTION_ON_OUTPUT_NODE";
 
-impl From<RendererInitError> for PipelineError {
-    fn from(err: RendererInitError) -> Self {
-        PipelineError {
-            error_code: "UNKNOWN_INIT_ERROR",
-            error_type: ErrorType::ServerError,
-            message: err.to_string(),
+impl From<UpdateSceneError> for PipelineError {
+    fn from(err: UpdateSceneError) -> Self {
+        match err {
+            UpdateSceneError::CreateNodeError(_, _) => {
+                PipelineError::new(FAILED_TO_CREATE_NODE, err, ErrorType::UserError)
+            }
+            UpdateSceneError::InvalidSpec(_) => {
+                PipelineError::new(SCENE_SPEC_VALIDATION_ERROR, err, ErrorType::UserError)
+            }
+            UpdateSceneError::NoNodeWithIdError(_) => {
+                // ServerError because it should be validated is spec validation
+                PipelineError::new(MISSING_NODE_WITH_ID, err, ErrorType::ServerError)
+            }
+            UpdateSceneError::WgpuError(err) => err.into(),
+            UpdateSceneError::UnknownResolutionOnOutput(_) => PipelineError::new(
+                UNKNOWN_RESOLUTION_ON_OUTPUT_NODE,
+                err,
+                ErrorType::ServerError,
+            ),
```

Comment:
I think we also need a single-node prams validation (e.g. builtins validations, shaders params validation).
Keep in mind, that sometimes to determine if specs are valid we need to know the input count (e.g. when param specifies how single input should be rendered)